### ISSUE TITLE
Thoughts on choosing the number of PCs

### ISSUE THREAD

**Issue by LTLA** [kevinblighe/PCAtools#5]
I've got a real nice implementation of Horn's parallel analysis in _scran_ that's looking for a better home.... would you like it? I don't use it anymore but it'd be a shame to throw it away.

It might also be a good idea to write a quick and dirty method for detecting the elbow point, given a vector of variance explained; if you haven't done so already. 

```r
# Mocking up some variance explained.
var.exp <- sort(rexp(100), decreasing=TRUE)

# Computing the elbow based on distance from diagonal.
dy <- -diff(range(var.exp))
dx <- length(var.exp) - 1
l2 <- sqrt(dx^2 + dy^2)
dx <- dx/l2
dy <- dy/l2

dy0 <- var.exp - var.exp[1]
dx0 <- seq_along(var.exp) - 1

parallel.l2 <- sqrt((dx0 * dx)^2 + (dy0 * dy)^2)
normal.x <- dx0 - dx * parallel.l2
normal.y <- dy0 - dy * parallel.l2
normal.l2 <- sqrt(normal.x^2 + normal.y^2)

which.max(normal.l2)
```

**Comment by kevinblighe** [comment]
Hey Aaron, thank you for the suggestions. This was going to be a future development point in later versions. Currently, in the vignette, I just take first _n_ PCs that account for 80% variation. When you mentioned _scran_, I realised the importance of this in determining _n_ PCs for dimension reduction in scRNAseq. Many years ago, I used to use Kaiser's rule.

I have been and will struggle for time to make these changes (also referring to your other suggestion https://github.com/kevinblighe/PCAtools/issues/1), but am aiming for the next Bioc release. Feel free to make any pull requests if you have time yourself. This will also eventually be published and sig. contributors will be 'rewarded'! I believe it will be a mainstay in Bioc in the long term.
